flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/StateChangelogStorageLoader.java

```diff
@@ -28,16 +34,51 @@
 /** A thin wrapper around {@link PluginManager} to load {@link StateChangelogStorage}. */
 @Internal
 public class StateChangelogStorageLoader {
-    private final PluginManager pluginManager;
 
-    public StateChangelogStorageLoader(PluginManager pluginManager) {
-        this.pluginManager = pluginManager;
+    private static final Logger LOG = LoggerFactory.getLogger(StateChangelogStorageLoader.class);
+
+    /**
+     * Mapping of state changelog storage identifier to the corresponding storage factories,
+     * populated in {@link StateChangelogStorageLoader#initialize(PluginManager)}.
+     */
+    private static final HashMap<String, StateChangelogStorageFactory>
+            STATE_CHANGELOG_STORAGE_FACTORIES = new HashMap<>();
+
+    static {
+        // Guarantee to trigger once.
+        initialize(null);
     }
 
-    @SuppressWarnings({"rawtypes"})
-    public Iterator<StateChangelogStorage> load() {
-        return concat(
-                pluginManager.load(StateChangelogStorage.class),
-                ServiceLoader.load(StateChangelogStorage.class).iterator());
+    public static void initialize(PluginManager pluginManager) {
+        STATE_CHANGELOG_STORAGE_FACTORIES.clear();
+        Iterator<StateChangelogStorageFactory> iterator =
+                pluginManager == null
+                        ? ServiceLoader.load(StateChangelogStorageFactory.class).iterator()
+                        : concat(
+                                pluginManager.load(StateChangelogStorageFactory.class),
+                                ServiceLoader.load(StateChangelogStorageFactory.class).iterator());
+        iterator.forEachRemaining(
+                factory ->
+                        STATE_CHANGELOG_STORAGE_FACTORIES.putIfAbsent(
+                                factory.getIdentifier().toLowerCase(), factory));
```

Comment:
Should we warn or fail if we have duplicated factories?
(given that we don't allow to initialize twice)

Reply:
We do ```clear()``` before load and ```putIfAbsent```, so initializing twice won't cause duplicated factories. Duplicated factories occur when ```PluginManager``` produces several factories with same name or ```ServiceLoader``` loads one with same name. This may happen when users put several jars in ```lib```.
Since JVM will load the first class from the first jar it finds, and the later ones with same identifiers are ignored, we should do the same. However, I do agree we remind user with a WARN message.